union-type.js

```diff
@@ -1,78 +1,130 @@
+    var type = typeof value;
+    return !!value && (type == 'object' || type == 'function');
+  };
+  var isFunction = function(f) { return typeof f === 'function'; };
+  var isArray = Array.isArray || function(a) { return 'length' in a; };
+
+  var mapConstrToFn = function(group, constr) {
+    return constr === String    ? isString
+         : constr === Number    ? isNumber
+         : constr === Boolean   ? isBoolean
+         : constr === Object    ? isObject
+         : constr === Array     ? isArray
+         : constr === Function  ? isFunction
+         : constr === undefined ? group
+                                : constr;
+  };
 
-var mapConstrToFn = function(group, constr) {
-  return constr === String    ? isString
-       : constr === Number    ? isNumber
-       : constr === Boolean   ? isBoolean
-       : constr === Object    ? isObject
-       : constr === Array     ? isArray
-       : constr === Function  ? isFunction
-       : constr === undefined ? group
-                              : constr;
-};
+  var numToStr = ['first', 'second', 'third', 'fourth', 'fifth', 'sixth', 'seventh', 'eighth', 'ninth', 'tenth'];
 
-function Constructor(group, name, validators) {
-  return curryN(validators.length, function() {
-    var val = [], validator, i, v;
-    if (Type.disableChecking === true) {
-      for (i = 0; i < arguments.length; ++i) val[i] = arguments[i];
-    } else {
-      for (i = 0; i < arguments.length; ++i) {
-        v = arguments[i];
-        validator = mapConstrToFn(group, validators[i]);
-        if ((typeof validator === 'function' && validator(v)) ||
-            (v !== undefined && v !== null && v.of === validator)) {
-          val[i] = v;
-        } else {
-          throw new TypeError('wrong value ' + v + ' passed to location ' + i + ' in ' + name);
-        }
+  var validate = function(group, validators, name, args) {
+    var validator, v, i;
+    for (i = 0; i < args.length; ++i) {
+      v = args[i];
+      validator = mapConstrToFn(group, validators[i]);
+      if (process.env.NODE_ENV !== 'production' &&
+	  (validator.prototype === undefined || !validator.prototype.isPrototypeOf(v)) &&
+          (typeof validator !== 'function' || !validator(v))) {
+        throw new TypeError('wrong value ' + v + ' passed to location ' + numToStr[i] + ' in ' + name);
       }
     }
-    val.of = group;
-    val.name = name;
-    return val;
-  });
+  };
+}
+
+function valueToArray(value) {
+  var i, arr = [];
+  for (i = 0; i < value.keys.length; ++i) {
+    arr.push(value[value.keys[i]]);
+  }
+  return arr;
 }
 
-function rawCase(type, cases, action, arg) {
-  if (type !== action.of) throw new TypeError('wrong type passed to case');
-  var name = action.name in cases ? action.name
-           : '_' in cases         ? '_'
-                                  : undefined;
-  if (name === undefined) {
-    throw new Error('unhandled value passed to case');
+function extractValues(keys, obj) {
```

Comment:
since we are not using es6:
```js
function extractValues (keys, obj) {
  return keys.map(function (x, index) {
    return obj[keys[index]];
 });
};
```
Admittedly not as nice ;) Maybe we could add es6 transpiling at some point ;)